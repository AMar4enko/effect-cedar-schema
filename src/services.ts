import { GenericTag } from "effect/Context"

export interface CedarSchema {
  readonly defaultNamespace: string
  readonly namespace: Map<string, {
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    entityTypes: Map<string, Record<string, any>>
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    commonTypes: Map<string, Record<string, any>>
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    actions: Map<string, Record<string, any>>  
  }>
}

export const CedarSchema = GenericTag<CedarSchema>(`@cedar-schema/schema`)
