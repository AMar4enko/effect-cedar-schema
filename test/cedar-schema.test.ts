import { Console, Effect } from 'effect'
import { compileFromActions } from '../src/index.ts'
import { expect, it } from '@effect/vitest'
import { CedarSchema } from '../src/services.ts'
import * as actions from './fixtures/actions.ts'

it.effect('compiles BookStore schema', () => Effect.gen(function* () {
  const schema = yield* compileFromActions([
    actions.CancelOrder,
    actions.PlaceOrder
  ]).pipe(
    Effect.tapErrorTag(`UnsupportedSchema`, (err) => 
      Console.error(err.toJSON())
    )
  )

  expect(schema).toMatchSnapshot()

}).pipe(
  Effect.provideService(CedarSchema, { defaultNamespace: `BookStore`, namespace: new Map() })
))