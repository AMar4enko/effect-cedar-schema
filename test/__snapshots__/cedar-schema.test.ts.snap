// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiles BookStore schema 1`] = `
{
  "BookStore": {
    "actions": {
      "cancelOrder": {
        "appliesTo": {
          "context": {
            "attributes": {
              "reason": {
                "required": true,
                "type": "String",
              },
            },
            "type": "Record",
          },
          "principalTypes": [
            "BookStore::User",
          ],
          "resourceTypes": [
            "BookStore::Order",
          ],
        },
      },
      "placeOrder": {
        "appliesTo": {
          "context": undefined,
          "principalTypes": [
            "BookStore::User",
          ],
          "resourceTypes": [
            "BookStore::Order",
          ],
        },
      },
    },
    "commonTypes": {
      "Stock": {
        "attributes": {
          "available": {
            "required": true,
            "type": "Long",
          },
          "reserved": {
            "required": true,
            "type": "Long",
          },
        },
        "type": "Record",
      },
    },
    "entityTypes": {
      "Author": {
        "shape": {
          "attributes": {
            "name": {
              "required": true,
              "type": "String",
            },
          },
          "type": "Record",
        },
      },
      "Book": {
        "shape": {
          "attributes": {
            "ageFrom": {
              "required": true,
              "type": "Long",
            },
            "author": {
              "name": "BookStore::Author",
              "required": true,
              "type": "Entity",
            },
            "title": {
              "required": true,
              "type": "String",
            },
          },
          "type": "Record",
        },
      },
      "Department": {
        "shape": {
          "attributes": {},
          "type": "Record",
        },
      },
      "Order": {
        "shape": {
          "attributes": {
            "books": {
              "element": {
                "name": "BookStore::Book",
                "type": "Entity",
              },
              "type": "Set",
            },
            "manager": {
              "name": "BookStore::User",
              "required": true,
              "type": "Entity",
            },
            "stock": {
              "type": "Stock",
            },
            "storage": {
              "name": "BookStore::Storage",
              "required": true,
              "type": "Entity",
            },
          },
          "type": "Record",
        },
      },
      "Role": {
        "shape": {
          "attributes": {
            "name": {
              "required": true,
              "type": "String",
            },
          },
          "type": "Record",
        },
      },
      "Storage": {
        "shape": {
          "attributes": {
            "location": {
              "required": true,
              "type": "String",
            },
          },
          "type": "Record",
        },
      },
      "User": {
        "memberOfTypes": [
          "BookStore::Role",
          "BookStore::Department",
        ],
        "shape": {
          "attributes": {
            "name": {
              "required": true,
              "type": "String",
            },
          },
          "type": "Record",
        },
      },
    },
  },
}
`;
